### DISCUSSION TITLE
Interaction between rw1 and categorical covariate in inlabru

### DISCUSSION THREAD

**Discussion by rita86ribeiro** [inlabru-org/inlabru#173] [Q&A] [answered]
Hi Finn, 

I am currently working on a LGCP model to predict the spatial distribution of bat roosts. As a first step, I have an observation model that describes roost density as a flexible function of accessibility (modelled as a rw1) and region (categorical covariate).  I would like to include an interaction term between the accessibility layer and region, but I am not sure how to do this in inlabru when I have a rw1. 

What I did so far is explained below. This runs ok and the effect of the non-linear effect of the accessibility seems realistic when adding the interaction with region, but I am not sure if what I did is correct.  I am just including the code, not uploading any data. But if you would like to run the example, please let me know. 

Thank you so much.
Rita Ribeiro

``` r
##layer 1 and 2 are spatial pixels data frames and effort and region are the column names of each one, respectively  

comp <-  coordinates ~
   Intercept(1) +
   effort(layer1, model="rw1", hyper = prec.prior) +
   region(layer2, model = "factor_contrast") ##define model components 

fit <- lgcp(comp,
             data = vbr,
             samplers = valleys,
             domain = list(coordinates = mesh),
             formula = coordinates ~ Intercept + effort + region + (effort*region), 
             options = list(control.compute = list(dic = TRUE, waic = TRUE, config = TRUE, cpo = TRUE),
                            control.inla = list(strategy = 'simplified.laplace', int.strategy="eb"),
                            verbose = TRUE))
```

**Comment by finnlindgren** [comment, marked as answer]
It helps to spell out the mathematical model definition. Your formula defines
```
  eta = beta_0 + fun(effort) + beta_region + fun(effort) * beta_region
```
where fun(effort) is exactly the same thing both times it appears, which is probably not what you want, as that's a direct functional interaction.
You probably want a covariance interaction, where you can have a different functional form of the effort effect for each region? E.g.
```
  eta = beta_0 + fun(effort) + beta_region + fun_region(effort)
```
or
```
  eta = beta_0 + beta_region + fun_region(effort)
```
For that you can use a replicated rw1 model; the same covariance parameters for each function, but different realisations.
For the latter option above, 
``` r
comp <-  ~
   Intercept(1) +
   effort(layer1, model="rw1", hyper = prec.prior, replicate = layer2, replicate_layer = 1, constr = TRUE) +
   region(layer2, model = "factor_contrast")

formula = coordinates ~ Intercept + effort + region   
```
If you want a separate "main effect", you need
``` r
comp <-  ~
   Intercept(1) +
   effort(layer1, model="rw1", hyper = prec.prior, constr = TRUE) +
   region(layer2, model = "factor_contrast") +
   effort_region(layer1, model="rw1", hyper = prec.prior, replicate = layer2, replicate_layer = 1, constr = TRUE)

formula = coordinates ~ Intercept + effort + region + effort_region
```
but this is non-identifiable model and may not work as intended.  It's likely easier to use the other approach, and if you then want the "common/main effect", calculate it as the average effect across the regions. One can in principle handle it through a lot of linear constraints instead, but that would be computationally expensive.

**Reply by finnlindgren** [reply]
In case you actually _do_ want a simple direct scaling of a single rw1 model, but with saling different from the region intercepts, then you need for examples this:
``` r
comp <-  ~
   Intercept(1) +
   effort(layer1, model="rw1", hyper = prec.prior) +
   region(layer2, model = "factor_contrast")
   region_effort_scaling(layer2, model = "factor_full")

formula = coordinates ~ Intercept + region + effort * region_effort_scaling
```
In this particular case, I would probably remove the intercept component entirely, and use "factor_full" also for the region intercepts.

Note that the inlabru "formula" input to inlabru is called that for historical and backwards compatibility reasons only; its right hand side is always a general R expression.  To express basic lm() type "interactions" that only involve "fixed effects", one can now define model components that use regular lm() formulas:
``` r
components = ~ ... + interactions(~ -1 + region + region:something, model="fixed")
```
Internally, that formula is given to `MatrixModels::modelMatrix()` to construct the component design matrix.

**Reply by finnlindgren** [reply]
Note: in the version with `effort * region_effort_scaling`, one would really want to fix the precision scaling parameter for `effort`. There may also be other identifiability issues...

**Comment by rita86ribeiro** [comment]
Finn, this is great. Thank you so much for all your help. I am going to explore these possibilities.

**Comment by rita86ribeiro** [comment]
Hi Finn, 

Regarding the discussion above, I still have some questions. I suspect that is something related with my data. 
When I run the first (or the second) solution you suggested and define replicate = layer2 (where layer2 is a spatial object from the class SpatialPixelsDataFrame), I get an error: 
Error in inla.core(formula = formula, family = family, contrasts = contrasts,  : 
#Error in the values of `replicate'; not in [1,...,2]

What I did was to substitute ‘layer2’ by the name of the column with data on the spatialpixelsdataframe (region) – and this runs without any warning or error. So: replicate = region 
Is this the possible cause of the error described above, and so, the solution I provide is correct? Or do this require a detailed analysis of the issue? 


> sessionInfo()
R version 4.1.2 (2021-11-01)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Monterey 12.0

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] rgeos_0.5-9        readxl_1.4.0       dplyr_1.0.8        ggpolypath_0.1.0   ggpubr_0.4.0       ggplot2_3.3.6     
 [7] RColorBrewer_1.1-3 rgdal_1.5-32       INLA_22.04.09      foreach_1.5.2      Matrix_1.5-1       inlabru_2.6.0.9003
[13] raster_3.5-15      sp_1.5-0          

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.9          lattice_0.20-45     tidyr_1.2.0         assertthat_0.2.1    digest_0.6.29       utf8_1.2.2         
 [7] R6_2.5.1            cellranger_1.1.0    plyr_1.8.7          backports_1.4.1     MatrixModels_0.5-1  stats4_4.1.2       
[13] evaluate_0.15       pillar_1.8.1        rlang_1.0.6         rstudioapi_0.13     car_3.0-12          rmarkdown_2.13     
[19] labeling_0.4.2      splines_4.1.2       munsell_0.5.0       broom_0.7.12        numDeriv_2016.8-1.1 compiler_4.1.2     
[25] xfun_0.30           pkgconfig_2.0.3     mnormt_2.0.2        tmvnsim_1.0-2       htmltools_0.5.2     tidyselect_1.1.2   
[31] tibble_3.1.8        gridExtra_2.3       codetools_0.2-18    fansi_1.0.3         viridisLite_0.4.1   withr_2.5.0        
[37] grid_4.1.2          gtable_0.3.1        lifecycle_1.0.3     DBI_1.1.2           magrittr_2.0.3      scales_1.2.1       
[43] cli_3.4.1           carData_3.0-5       farver_2.1.1        ggsignif_0.6.3      sn_2.0.2            generics_0.1.2     
[49] vctrs_0.4.2         iterators_1.0.14    tools_4.1.2         glue_1.6.2          purrr_0.3.4         abind_1.4-5        
[55] fastmap_1.1.0       yaml_2.3.5          colorspace_2.0-3    terra_1.5-21        rstatix_0.7.0       knitr_1.38         
[61] patchwork_1.1.2

**Comment by finnlindgren** [comment]
Your workaround sounds wrong, and possibly an indication of a bug/misfeature that doesn’t notice.

For the replicate information to be correct (not giving an error doesn’t mean it’s correct) it needs to produce a vector of positive integers.  When the replicate object is given as a spdf, it should normally evaluate the first layer of that spdf. To be clear about which element to extract, specify the replicate_layer as an integer or string indicating which column/layer to use.

to test it, call eval_spatial directly and see what comes out.

**Reply by rita86ribeiro** [reply]
Finn, thank you so much. I will test.

**Comment by rita86ribeiro** [comment]
Hi Finn, 

When calling `eval_spatial(layer2, points, layer = 1)` I get the corresponding value of my layer2 for each point evaluated.
So, what I did was to run the same expression: 

`comp <-  ~ Intercept(1) + effort(layer1, model="rw1", hyper = prec.prior, replicate = layer2, replicate_layer = 1, constr = TRUE) + region(layer2, model = "factor_contrast")`
`formula = coordinates ~ Intercept + effort + region`

with another covariate instead of my layer2. The model run as expected. So, it seems to be an issue with my layer2. This layer is a factor covariate, which defines 2 regions "Apurimac" and "AyaCus". The other layer I used for this test is not a factor. 

I tested the same with the gorillas dataset and the vegetation layer, and I obtained the same error 
"Error in inla.core(formula = formula, family = family, contrasts = contrasts,  : 
  Error in the values of `replicate'; not in [1,...,6]" 

Where 1-6 are the levels of the vegetation covariate. I don't know if this is a bug or if I am missing something on these models. 

Here is the code with the gorillas dataset: 
`library(INLA)`
`library(inlabru)`
`library(RColorBrewer)`
`library(ggplot2)`

`bru_options_set(inla.mode = "experimental")`
`data(gorillas, package = "inlabru")`
`nests <- gorillas$nests`
`mesh <- gorillas$mesh`
`boundary <- gorillas$boundary`
`gcov <- gorillas$gcov`

 ##I used the covariate waterdist to be similar to my layer2 (accessibility = time travelling)


`prec.prior <- list(prec = list(prior="pc.prec", param = c(0.1, 0.01)))`

`comp1 <- coordinates ~ Intercept(1) + waterdist(gcov$waterdist, model="rw1", hyper = prec.prior, replicate = gcov$vegetation, replicate_layer = 1, constr = TRUE) + vegetation(gcov$vegetation, model = "factor_contrast")`


`fit1 <- lgcp(comp1, nests, samplers = boundary, domain = list(coordinates = mesh), options = list(control.compute = list(dic = TRUE, waic = TRUE, config = TRUE, cpo = TRUE),  control.inla = list(strategy = 'simplified.laplace', int.strategy="eb"), verbose = TRUE))`

Error in inla.core(formula = formula, family = family, contrasts = contrasts,  : 
  Error in the values of `replicate'; not in [1,...,6]

 *** inla.core.safe:  inla.program has crashed: rerun to get better initial values. try=1/2 
Error in inla.core(formula = formula, family = family, contrasts = contrasts,  : 
  Error in the values of `replicate'; not in [1,...,6]

 *** inla.core.safe:  inla.program has crashed: rerun to get better initial values. try=2/2 
Error in inla.core(formula = formula, family = family, contrasts = contrasts,  : 
  Error in the values of `replicate'; not in [1,...,6]
Warning in iinla(model = info[["model"]], lhoods = info[["lhoods"]], options = info[["options"]]) :
  iinla: Problem in inla: Error in inla.core.safe(formula = formula, family = family, contrasts = contrasts,  : 
  *** Fail to get good enough initial values. Maybe it is due to something else.
iinla: Problem in inla: 1: lgcp(comp1, nests, samplers = boundary, domain = list(coordinates = mesh), 
       options = list(control.compute = list(dic = TRUE, waic = TRUE, 
           config = TRUE, cpo = TRUE), control.inla = list(strategy = "simpli [...]
           int.strategy = "eb"), verbose = TRUE))
2: bru(components, lik, options = options)
3: iinla(model = info[["model"]], lhoods = info[["lhoods"]], options = info[[ [...]
4: try_callstack(...)
5: do.call(INLA::inla, inla.options.merged, envir = environment(model$effects))
6: (function (formula = NULL, family = "gaussian", contrasts = NULL, 
       data = NULL, quantiles = c(0.025, 0.5, 0.975), E = NULL, 
       offset = NULL, scale = NULL, weights = NULL, Ntrials = NULL, 
       strata = NULL, lp.scale = NULL, link.covariates = NULL, verb [... truncated]
iinla: Giving up and returning last successfully obtained result for diagnostic purposes.



Thank you

**Reply by finnlindgren** [reply]
Looks like the default mapper construction needs to detect factors for `replicate` and use a different mapper class than it does now, since it doesn't seem to like the mapper currently used for factor `replicate` input. I think one of the existing classes does what is needed.
Can you show the output of `summary(bru_model(components, like_list(like(formula=coordinate~., model="cp", ...))))`? (the `like()` call should have the inputs you give to `lgcp()`, but without the components)

Until I figure out a general fix, you can work around it by doing the conversion yourself:
```r
+ effort(layer1, model="rw1", hyper = prec.prior, constr = TRUE,
         replicate = as.numeric(eval_spatial(layer2, .data., layer = 1)))
```

**Reply by finnlindgren** [reply]
Tracking this as issue #174

**Reply by rita86ribeiro** [reply]
Hi Finn, 
Thank you very much. 

Hum, when I compute `summary(bru_model(comp1, like_list(like(formula = coordinates ~ Intercept + effort + Region,
                                        model="cp", 
                                        data=vbr))))` 

I get an error: Error in (function (cond)  : 
  error in evaluating the argument 'object' in selecting a method for function 'summary': inherits(components, "component_list") is not TRUE

**Reply by finnlindgren** [reply]
Should probably be this:
```r
summary(bru_model(
  component_list(comp1),
  like_list(like(formula = coordinates ~ Intercept + effort + Region, model="cp", data=vbr))
))
```

**Reply by finnlindgren** [reply]
...but you also need to supply the `samplers` and `domain` information to `like()`.

**Reply by rita86ribeiro** [reply]
Hi Finn, 

Thank you, so: 
`summary(bru_model(
component_list(comp1), 
like_list(like(formula = coordinates ~ Intercept + effort + Region, data=vbr, samplers = expval, domain = list(coordinates = mesh1), family = "cp"))))` 

gives: 
        Length Class          Mode
effects 3      component_list list
formula 3      formula        call

**Reply by finnlindgren** [reply]
Ah, sorry, I thought I had fixed that summary function…  I need summary(bru_model(…)$effects)

**Reply by rita86ribeiro** [reply]
Hi Finn, 

Sorry..I was not familiar with this summary function.
Here is the output: 

```
summary(bru_model(component_list(comp1), like_list(like(formula = coordinates ~ Intercept + effort + Region, data=vbr, samplers = expval, domain = list(coordinates = mesh1), family = "cp")))$effects)

Label:	Intercept
  Main:	type 'linear', input '1'
  Map: pipe = multi(main = linear, group = index, replicate = index) -> scale
  INLA formula:	
    ~ . + f(Intercept, model = BRU_Intercept_main_model, ngroup = 1, nrep = 1, values = BRU_Intercept_values)
Label:	effort
  Main:	type 'rw1', input 'acces_senasa'
  Replicate:	type 'iid', input 'aacf'
  Map: pipe = multi(main = inla_mesh_1d, group = index, replicate = factor) -> scale
  INLA formula:	
    ~ . + f(effort, model = BRU_effort_main_model, hyper = prec.prior, replicate = effort.repl, constr = TRUE,
      ngroup = 1, nrep = 2, values = BRU_effort_values)
Label:	Region
  Main:	type 'factor', input 'aacf'
  Map: pipe = multi(main = factor, group = index, replicate = index) -> scale
  INLA formula:	
    ~ . + f(Region, model = BRU_Region_main_model, hyper = BRU_Region_main_fixed_hyper, ngroup = 1, nrep = 1, values
      = BRU_Region_values)
```

Thank you

**Reply by finnlindgren** [reply]
Thanks! The summary functions for the bru objects aren’t complete yet; in the future it will be easier to get to this information…

what I see is that some of the replicate mappers come out as factor mappers, which doesn’t do what we want in these cases. It’s the right thing for the main inout and also I think for group, but replicate seems to have different internal inla logic that inlabru needs to take into account. I’ll look into how to best make it work by default.

**Reply by rita86ribeiro** [reply]
Thank you so much Finn.

**Reply by finnlindgren** [reply]
I believe this has now been fixed in the development version on github. Both group and replicate mappers are now by default using integer mappings in the `INLA::f()` calls even when the inputs are given as factors.

**Reply by rita86ribeiro** [reply]
This is great Finn. Thank you. I will test with my data.

**Reply by finnlindgren** [reply]
The automated test indicate something isn't right with the fix, so keep using the workaround for now.

**Reply by finnlindgren** [reply]
I had missed uploading the actual fix part of the fix. I think it will work now.
There's another issue I just noticed where the new rgdal version gives lots of deprecation warning when using `sp::spTransform`; sp was supposed to keep working when rdal goes away sometime next year, so that looks like a problem with sp, but I'll likely have to work around it in our code anyway. We are in the process of adding sf support, but I really don't want sp support to stop working.

**Reply by rita86ribeiro** [reply]
Hi Finn, 

Ok, so with this fix the expression `effort(layer1, model="rw1", hyper = prec.prior, replicate = layer2, replicate_layer = 1, constr = TRUE)`, where layer2 is a factor, should run ok right? 
Thank you

**Reply by finnlindgren** [reply]
When "the first layer of layer2 is a factor", yes.

**Reply by rita86ribeiro** [reply]
Yes, it runs perfect now. 
Thank you so much for all this help.
